### PR TITLE
Shutdown RPC endpoint

### LINKED ISSUE #2402: Implement forest-cli daemon stop command.
**Issue summary**

Forest is capable of shut-downing itself because it installed a handler that catches `SIGINT` and `SIGTERM` signals.
So if forest is running in the foreground, hitting Ctrl+C will stop the node.

We would like an equivalent of the Shutdown endpoint in common api of Lotus, as an alternative of resorting only on `ctrl+c` or `kill` commands.

**Task summary**

- [x] Expose a `Shutdown` RPC endpoint with admin rights
- [x] Expose some internal signal to gracefully shutdown forest
- [x] Bind `ctrc+c` and RPC inbound message to it

**Acceptance Criteria**

- [x] Calling `forest-cli shutdown` shutdowns the node
- [x] Using signals (ie `ctrl+c`) still works

**Other information and links**

https://lotus.filecoin.io/lotus/manage/lotus-cli/#lotus-daemon-stop
https://github.com/filecoin-project/lotus/blob/master/api/api_common.go#L51

Note that this command will require that the RPC server be running.

### PR DESCRIPTION
**Pull request by elmattic**
## Summary of changes

Changes introduced in this pull request:
- Add `Filecoin.Shutdown` endpoint and `forest-cli shutdown` subcommand
- Expose `version()` and `shutdown()` in js console
- Add `sleep` and `sleepTipsets` builtins in js console
- Fix bug in module path resolver

Will rebase on main once js-console branch is merged.

## Reference issue to close (if applicable)

Closes #2402

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

**Comment by elmattic** [comment]
Forgot to add the `forest-cli shutdown` subcommand.

File: forest/daemon/src/daemon.rs [hunk 1]
```diff
@@ -314,18 +318,24 @@ pub(super) async fn start(config: Config, detached: bool) -> anyhow::Result<Db>
...
 
-    select! {
+    tokio::select! {
         () = sync_from_snapshot(&config, &state_manager).fuse() => {},
-        _ = ctrlc_oneshot => {
-            // Cancel all async services
+        _ = tokio::signal::ctrl_c() => {
+            services.shutdown().await;
+            return Ok(db);
+        },
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:326]: Just to make sure, this way of handling ctrl-c still maintains the same behavior of attempting a soft shutdown on the first ctrl-c and then doing a hard shutdown on subsequent ctrl-c's?

**Comment by lemmih** [comment]
What's the purpose of `sleep` and `sleepTipsets`?

File: forest/daemon/src/daemon.rs (hunk 1 above)
Comment (elmattic) [reply forest/daemon/src/daemon.rs:326]: I don't think so, I will confirm it to you. Do you think it's still needed?

**Comment by elmattic** [comment]
> What's the purpose of `sleep` and `sleepTipsets`?

Let's say you want to send some tokens and once message is mined, do something else. You can use sleepTipsets for that purpose, ie:

```
> showWallet()
Address                                         Balance
f1am5w6zocm5t2weizopnvimyww3y7t4oqmev2bma       999999697339093950 attoFIL
f1hxovphj7nheqo35u732rebudd4zvfzskwjwjdza       0 attoFIL
f1tcbi3liho447tbsvkqtodpnjpemnvpuovxdd5ta       100000 attoFIL
f1w2v5zoguqevqffmzmaieveosfuskz7d7t5mdw6a       100000000000000000000 attoFIL
f1w5jzluupzcvvj72yoek5ms3phodlispghykc7ni       1000000000000000000 attoFIL

> sendFIL("f1tcbi3liho447tbsvkqtodpnjpemnvpuovxdd5ta", 200000); sleepTipsets(2); showWallet()
Address                                         Balance
f1am5w6zocm5t2weizopnvimyww3y7t4oqmev2bma       999999638065181910 attoFIL
f1hxovphj7nheqo35u732rebudd4zvfzskwjwjdza       0 attoFIL
f1tcbi3liho447tbsvkqtodpnjpemnvpuovxdd5ta       300000 attoFIL
f1w2v5zoguqevqffmzmaieveosfuskz7d7t5mdw6a       100000000000000000000 attoFIL
f1w5jzluupzcvvj72yoek5ms3phodlispghykc7ni       1000000000000000000 attoFIL


```

File: forest/daemon/src/daemon.rs (hunk 1 above)
Comment (lemmih) [reply forest/daemon/src/daemon.rs:326]: Not sure. But if we change the behavior then we should make a note about it.

Comment (elmattic) [reply forest/daemon/src/daemon.rs:326]: Ok, to me we don't really need anymore to handle this hard exit. Forest should be able to gracefully shutdown now.

Comment (lemmih) [reply forest/daemon/src/daemon.rs:326]: Alright. Let's make a note of it in the changelog.

File: forest/cli/src/cli/shutdown_cmd.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+use forest_rpc_client::common_shutdown;
+
+use super::{handle_rpc_err, Config};
+use crate::cli::prompt_confirm;
+
+#[derive(Debug, clap::Args)]
+pub struct ShutdownCommand {
+    /// Assume "yes" as answer to shutdown prompt
+    #[arg(long, short)]
+    yes: bool,
```
Comment (LesnyRumcajs) [inline forest/cli/src/cli/shutdown_cmd.rs:13]: Let's keep it to `force` like in
https://github.com/ChainSafe/forest/blob/main/forest/cli/src/cli/snapshot_cmd.rs#L93-L95

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
1. Let's cover this subcommand in tests, to replace the barbaric `pkill` https://github.com/ChainSafe/forest/blob/main/.github/workflows/rust.yml#L176
2. It's worth adding an entry in the changelog.

Does these make sense to you?

File: forest/cli/src/cli/shutdown_cmd.rs (hunk 2 above)
Comment (elmattic) [reply forest/cli/src/cli/shutdown_cmd.rs:13]: Sure, will do!

**Comment by elmattic** [comment]
> 1. Let's cover this subcommand in tests, to replace the barbaric `pkill` https://github.com/ChainSafe/forest/blob/main/.github/workflows/rust.yml#L176
> 2. It's worth adding an entry in the changelog.
> 
> Does these make sense to you?

Will create an issue for point 1. Will be easier to do when calibnet checks are back.

File: forest/cli/src/cli/attach_cmd.rs [hunk 3]
```diff
@@ -236,6 +239,42 @@ async fn send_message(
...
+type SleepTipsetsResult = ();
+
+async fn sleep_tipsets(
+    params: SleepTipsetsParams,
+    auth_token: &Option<String>,
+) -> Result<SleepTipsetsResult, jsonrpc_v2::Error> {
+    let mut epoch = None;
+    loop {
+        let state = sync_status((), auth_token).await?;
+        if state.active_syncs[0].stage() == SyncStage::Complete {
```
Comment (creativcoder) [inline forest/cli/src/cli/attach_cmd.rs:264]: can `[0]` panic?

**Review by creativcoder — APPROVED** [review]

File: forest/cli/src/cli/attach_cmd.rs (hunk 3 above)
Comment (elmattic) [reply forest/cli/src/cli/attach_cmd.rs:264]: No. By construction the vector is constructed with one element.

https://github.com/ChainSafe/forest/blob/main/node/rpc/src/sync_api.rs#L56

Comment (creativcoder) [reply forest/cli/src/cli/attach_cmd.rs:264]: :+1:

**Review by LesnyRumcajs — APPROVED** [review]

**Review by lemmih — CHANGES_REQUESTED** [review]
The cursor can go missing if we hit ctrl-c during a prompt.

**Review by hanabi1224 — APPROVED** [review]

**Review by lemmih — APPROVED** [review]

**Pull request merged**